api/src/main/java/org/eclipse/microprofile/config/ConfigValueResolver.java

```diff
@@ -0,0 +1,294 @@
+ * missing properties or failed conversion but instead return a default value that callers have to provide.
+ * Alternatively to providing a default value a value can be resolved as {@link Optional}.
+ *
+ * <p>
+ * Arrays of values can be resolved as {@link List} or {@link Set}.
+ * By default these return empty lists or sets in case of missing property or failed conversion.
+ *
+ * <p>
+ * Simple value properties defined as empty string by default are considered missing. This can be changed
+ * using {@link #acceptEmpty()}.
+ *
+ * <p>
+ * Should exceptions be thrown for either missing properties or failed conversion the default of not throwing exception
+ * and using default values can be overridden using {@link #throwOnMissingProperty()} and
+ * {@link #throwOnFailedConversion()}.
+ *
+ * <p>
+ * Usually conversion relies on registered {@link org.eclipse.microprofile.config.spi.Converter}s.
+ * Ad-hoc conversion can be done using {@link #asConvertedBy(Function, Object)}.
+ * Values resolved as {@link String} are not converted.
+ * Values resolved as {@code String[]} are only split but elements are not converted.
+ *
+ * <p>
+ * Typed defaults are provided with one of the {@code as}-methods.
+ * Raw property string defaults can be provided using {@link #withDefault(String)}.
+ * As the API can not capture if such a raw default has been provided it still requires to provide a typed default
+ * or use {@link Optional} to ensure resolution will return a value.
+ * If a raw {@link String} is provided it takes precedence over a typed default.
+ * Should conversion of a raw default fail the typed default is used
+ * or in case of {@link Optional} the property is considered not present.
+ *
+ * <p>
+ * The only way {@code as}-methods might return a {@code null} reference is by using {@code null} as typed default.
+ * In such case it is considered the intention of the caller and therefore not problematic.
+ *
+ * @author <a href="mailto:jan.bernitt@payara.fish">Jan Bernitt</a>
+ */
+public interface ConfigValueResolver {
+
+    /**
+     * Adds a profile to property resolution which is applied when requesting a value.
+     * Multiple calls to this method on the same instance override prior profile values.
+     *
+     * @param value name of the profile to use, {@code null} is ignored
+     * @return This resolver for fluent API usage
+     */
+    ConfigValueResolver withProfile(String value);
+
+    /**
+     * Provides a raw property value default value.
+     * <p>
+     * The raw default is considered to be on the source level and therefore takes precedence over typed defaults
+     * provided to any of the {@code as}-methods. Should raw default be empty of fail to convert to the target type the
+     * typed default is used.
+     *
+     * @param value raw default value, not {@code null}
+     * @return This resolver for fluent API usage
+     */
+    ConfigValueResolver withDefault(String value);
```

Comment:
It is my view that having non-global default values will only add a layer of confusion.  In Quarkus, users were expecting that default values be included in property expression resolution, and we had to retrofit in order to support that.  This just reintroduces the problem again.  Default values should be configured when the configuration is being built, not at evaluation time (after all a configuration is defined to apply to a single application, so it should be possible to know what the default values would be at that stage).

Reply:
The best practices for mandating the global default is to specify the value in `microprofile-config.properties`. I am not sure whether all users give the same defaultValue on multiple property injections or not? Do you have any user feedback @dmlloyd on this?

Reply:
It is required in Quarkus. Nobody has complained.